test/parallel/test-repl-history-navigation.js

```diff
@@ -0,0 +1,128 @@
+'use strict';
+
+// Flags: --expose-internals
+
+const common = require('../common');
+const stream = require('stream');
+const REPL = require('internal/repl');
+const assert = require('assert');
+
+// Create an input stream specialized for testing an array of actions
+class ActionStream extends stream.Stream {
+  run(data) {
+    const _iter = data[Symbol.iterator]();
+    const doAction = () => {
+      const next = _iter.next();
+      if (next.done) {
+        // Close the repl. Note that it must have a clean prompt to do so.
+        setImmediate(() => {
+          this.emit('keypress', '', { ctrl: true, name: 'd' });
+        });
+        return;
+      }
+      const action = next.value;
+
+      if (typeof action === 'object') {
+        this.emit('keypress', '', action);
+      } else {
+        this.emit('data', `${action}`);
+      }
+      setImmediate(doAction);
+    };
+    setImmediate(doAction);
+  }
+  resume() {}
+  pause() {}
+}
+ActionStream.prototype.readable = true;
+
+
+// Mock keys
+const ENTER = { name: 'enter' };
+const UP = { name: 'up' };
+const DOWN = { name: 'down' };
+
+const prompt = '> ';
+
+const tests = [
+  { // creates few history to navigate for
+    env: {},
+    test: [ 'let ab = 45', ENTER,
+            '555 + 909', ENTER,
+            '{key : {key2 :[] }}', ENTER],
+    expected: []
+  },
+  {
+    env: {},
+    test: [UP, UP, UP, DOWN, DOWN, DOWN],
+    expected: [prompt,
+               `${prompt}{key : {key2 :[] }}`,
+               `${prompt}555 + 909`,
+               `${prompt}let ab = 45`,
+               `${prompt}555 + 909`,
+               `${prompt}{key : {key2 :[] }}`,
+               `${prompt}`]
+  }
+];
+const numtests = tests.length;
+
+const runTestWrap = common.mustCall(runTest, numtests);
+
+function runTest() {
+  const opts = tests.shift();
+  if (!opts) return; // All done
+
+  const env = opts.env;
+  const test = opts.test;
+  const expected = opts.expected;
+
+  REPL.createInternalRepl(env, {
+    input: new ActionStream(),
+    output: new stream.Writable({
+      write(chunk, _, next) {
+        const output = chunk.toString();
+
+        if (output.charCodeAt(0) === 27 || /^[\r\n]+$/.test(output)) {
+          return next();
+        }
+
+        if (expected.length) {
+          const toBeAsserted = expected[0];
+          try {
+            assert.strictEqual(output, toBeAsserted);
+            expected.shift();
+          } catch (err) {
+            console.error(`Failed test # ${numtests - tests.length}`);
+            throw err;
+          }
+        }
+
+        next();
+      }
+    }),
+    prompt: prompt,
+    useColors: false,
+    terminal: true
+  }, function(err, repl) {
+    if (err) {
+      console.error(`Failed test # ${numtests - tests.length}`);
+      throw err;
+    }
+
+    repl.once('close', () => {
+      try {
+        // Ensure everything that we expected was output
+        assert.strictEqual(expected.length, 0);
```

Comment:
There should be no need to throwing an error that is then caught again. Instead, just test for the expectation right away:

```js
if (expected.length !== 0) {
  throw new Error(`Failed test # ${numtests - tests.length}`);
}
setImmediate(runTestWrap, true);
```

The same above.